packages/discovery/src/discovery/permission-resolving/resolvePermissions.test.ts

```diff
@@ -0,0 +1,818 @@
+        type: 'upgrade',
+        path: [
+          { address: contract1, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: contract2, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: l2Contract1, delay: 0 },
+          { address: l2Proxy, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: l2Contract2, delay: 0 },
+          { address: l2Proxy, delay: 0 },
+          { address: l1Proxy, delay: 0 },
+          { address: admins, delay: 0 },
+        ],
+      },
+    ])
+  })
+
+  it('zksync lite', () => {
+    const verifier = EthereumAddress.random()
+    const governance = EthereumAddress.random()
+    const zkSync = EthereumAddress.random()
+    const upgradeGatekeeper = EthereumAddress.random()
+    const zkSyncMsig = EthereumAddress.random()
+    const securityCouncil = EthereumAddress.random()
+
+    const graph: Graph = {
+      nodes: [
+        { address: verifier, members: [], threshold: 1 },
+        { address: governance, members: [], threshold: 1 },
+        { address: zkSync, members: [], threshold: 1 },
+        { address: upgradeGatekeeper, members: [], threshold: 1 },
+        { address: zkSyncMsig, members: [], threshold: 1 },
+        { address: securityCouncil, members: [], threshold: 1 },
+      ],
+      edges: [
+        { type: 'upgrade', fromNode: 0, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 1, toNode: 3, delay: 0 },
+        { type: 'upgrade', fromNode: 2, toNode: 3, delay: 0 },
+        { type: 'act', fromNode: 3, toNode: 4, delay: 21 },
+        { type: 'configure', fromNode: 2, toNode: 5, delay: 0 },
+      ],
+    }
+
+    expect(resolvePermissions(graph)).toEqualUnsorted([
+      {
+        type: 'upgrade',
+        path: [
+          { address: verifier, delay: 0 },
+          { address: upgradeGatekeeper, delay: 0 },
+          { address: zkSyncMsig, delay: 21 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: governance, delay: 0 },
+          { address: upgradeGatekeeper, delay: 0 },
+          { address: zkSyncMsig, delay: 21 },
+        ],
+      },
+      {
+        type: 'upgrade',
+        path: [
+          { address: zkSync, delay: 0 },
+          { address: upgradeGatekeeper, delay: 0 },
+          { address: zkSyncMsig, delay: 21 },
+        ],
+      },
+      {
+        type: 'configure',
+        path: [
+          { address: zkSync, delay: 0 },
+          { address: securityCouncil, delay: 0 },
+        ],
+      },
+    ])
+  })
+
+  it('one actor, four contracts, two timelocks with same delays', () => {
```

Comment:
Edge cases:
Two upgrades same delay, different path
Two upgrades different delay, different path
Configure longer delay than upgrade (collapsed?)
Configure same delay different path (collapsed or not?)